### PR TITLE
PADV 719 - Improve LTI launch view post method 

### PR DESCRIPTION
**Pull request by alexjmpb**
## Ticket

https://agile-jira.pearson.com/browse/PADV-719

## Description

This PR refactors the whole LTI launch view and its tests to encapsulate logic and reduce efforts when implementing new functionalities

## Type of Change

- [x] Refactor LtiToolLaunchView and create a method for each launch step.
- [x] Rewrite tests and create a testcase for each of the methods

## Testing:

- Run the LMS: `make dev.up.lms`.
- Install `openedx_lti_tool_plugin` on the LMS.
- Add required settings to LMS.
	
```
# Enable LTI Tool Provider Plugin.
OLTITP_ENABLE_LTI_TOOL = True
```

- Run ngrok or any other similar service: `ngrok http 18000`
- Create a new LTI 1.3 tool key config: http://localhost:18000/admin/lti1p3_tool_config/ltitoolkey/add/ (You can use IMS RI to create a key pair: https://lti-ri.imsglobal.org/keygen/index)
- Create a new LTI 1.3 tool config: http://localhost:18000/admin/lti1p3_tool_config/ltitool/add/

```
Issuer: https://saltire.lti.app/platform
Client id: saltire.lti.app
Auth login: url: https://saltire.lti.app/platform/auth
Auth token url: https://saltire.lti.app/platform/token/sda42bb0cf2352259a367a404d48d54e8
Key set url: https://saltire.lti.app/platform/jwks/sda42bb0cf2352259a367a404d48d54e8
Deployment ids: ["cLWwj9cbmkSrCNsckEFBmA"]
```

- Go to the saLTIre platform https://saltire.lti.app/platform
- Go to "Security Model" on the left sidebar and set these settings:

```
Message URL: https://{ngrok_url}/openedx_lti_tool_plugin/1.3/launch/{course_id}
Initiate login URL: https://{ngrok_url}/openedx_lti_tool_plugin/1.3/login
Redirection URI(s): https://{ngrok_url}/openedx_lti_tool_plugin/1.3/launch/{course_id}
Public keyset URL: https://{ngrok_url}/openedx_lti_tool_plugin/1.3/pub/jwks
```

- Click on the "Save" button on the top navbar.
- Click on the dropdown next to the "Connect" button on the top navbar.
- Click on "Open in iframe".
- The launch should work.
- Test that all the launch functionalities are working properly.

## Reviewers

- [x] @Squirrel18 
- [x] @kuipumu

### CODE REVIEW

File: openedx_lti_tool_plugin/views.py [hunk 1]
```diff
@@ -155,6 +156,68 @@ def post(self, request: HttpRequest) -> Union[HttpResponseRedirect, LoggedHttpRe
...
+        """
+        try:
+            return launch_message.get_launch_data()
+        except LtiException as exc:
+            raise LtiToolLaunchException(_(f'Launch message validation failed: {exc}')) from exc
+
+    def get_identity_claims(
+        self,
+        launch_data: dict
+    ) -> Tuple[str, Union[str, list], str, dict]:
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/views.py:179]:
```suggestion
    ) -> Tuple[str, Union[str, list], str, dict]:
```
What is the return `Union[str, list]` referring to?

File: openedx_lti_tool_plugin/views.py [hunk 2]
```diff
@@ -155,6 +156,68 @@ def post(self, request: HttpRequest) -> Union[HttpResponseRedirect, LoggedHttpRe
...
+            A tuple containing the iss, aud, sub and pii claims.
+        """
+        return (
+            launch_data.get('iss'),
+            get_client_id(launch_data.get('aud'), launch_data.get('azp')),
+            launch_data.get('sub'),
+            get_pii_from_claims(launch_data) if SAVE_PII_DATA.is_enabled() else {},
+        )
+
+    def check_course_access_permission(self, course_id: str, iss: str, aud: str) -> None:
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/views.py:195]: Functions without return don't need to be hinted

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 3]
```diff
@@ -0,0 +1,817 @@
...
+            get_identity_claims_mock: Mocked 'get_identity_claims' method.
+            get_launch_data_mock: Mocked 'get_launch_data' method.
+            message_launch_mock: Mocked 'DjangoMessageLaunch' class.
+        """
+        get_identity_claims_mock.return_value = (
+            ISS,
+            AUD,
+            SUB,
+            PII,
+        )
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:93]:
```suggestion
        get_identity_claims_mock.return_value = (ISS, AUD, SUB, PII)
```

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 4]
```diff
@@ -0,0 +1,817 @@
...
+        handle_ags_mock,
+        enroll_mock,  # pylint: disable=unused-argument
+        authenticate_and_login_mock,  # pylint: disable=unused-argument
+        get_lti_profile_mock,  # pylint: disable=unused-argument
+        check_course_access_permission_mock,  # pylint: disable=unused-argument
+        get_identity_claims_mock,
+        get_launch_data_mock,  # pylint: disable=unused-argument
+        message_launch_mock,
+    ):
+        """Test POST request when is not resource link.
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:200]:
```suggestion
        """Test POST request when LTI launch is not a resource link launch.
```

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 5]
```diff
@@ -0,0 +1,817 @@
...
+        message_launch_mock().is_resource_launch.assert_called_once_with()
+        get_resource_launch_mock.assert_called_once_with(COURSE_ID, USAGE_KEY)
+        handle_ags_mock.assert_called_once_with(
+            message_launch_mock(),
+            get_launch_data_mock(),
+            get_lti_profile_mock(),
+            USAGE_KEY
+        )
+
+    def test_post_no_resource_launch(
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:185]:
```suggestion
    def test_post_without_resource_link_launch(
```

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 6]
```diff
@@ -0,0 +1,817 @@
...
+        )
+        message_launch_mock().is_resource_launch.return_value = False
+        request = self.factory.post(self.url)
+        request.user = self.user
+
+        response = self.view_class.as_view()(request, COURSE_ID, USAGE_KEY)
+
+        get_resource_launch_mock.assert_not_called()
+        handle_ags_mock.assert_not_called()
+        self.assertEqual(response.status_code, 400)
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:217]: @alexjmpb Can you validate getext is called with the exception message?

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 7]
```diff
@@ -0,0 +1,802 @@
...
+        launch_message_mock.get_launch_data.assert_called_once_with()
+
+    def test_get_launch_data_raises_exception(self):
+        """Test LtiToolLaunchView get_launch_data method raises exception when LtiException is catched."""
+        launch_message_mock = MagicMock(
+            get_launch_data=MagicMock(side_effect=LtiException),
+        )
+
+        with self.assertRaises(LtiToolLaunchException):
+            self.view_class().get_launch_data(launch_message_mock)
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:238]: @alexjmpb Can you validate gettext is called with the exception message?

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 8]
```diff
@@ -0,0 +1,802 @@
...
+            .first.return_value
+        )
+
+        self.view_class().check_course_access_permission(COURSE_ID, ISS, AUD)
+
+        course_access_configuration_switch_mock.is_enabled.assert_called_once_with()
+        tool_config_mock.get_lti_tool.assert_not_called()
+        course_access_configuration_mock.objects.filter.assert_not_called()
+        course_access_configuration_mock.objects.filter.return_value.first.assert_not_called()
+        course_access_conf_queryset_mock.is_course_id_allowed.assert_not_called()
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:378]: @alexjmpb Can you validate gettext is called with the exception message?

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 9]
```diff
@@ -0,0 +1,802 @@
...
+            course_access_configuration_mock: Mocked CourseAccessConfiguration Model.
+            tool_config_mock: Mocked tool_config attribute of LtiToolLaunchView.
+            course_access_configuration_switch_mock: Mocked COURSE_ACCESS_CONFIGURATION waffle switch.
+        """
+        course_access_configuration_switch_mock.is_enabled.return_value = True
+        course_access_configuration_mock.objects.filter.return_value.first.return_value = None
+
+        with self.assertRaises(LtiToolLaunchException) as ex:
+            self.view_class().check_course_access_permission(COURSE_ID, ISS, AUD)
+        self.assertEqual(
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:398]: @alexjmpb Can you validate gettext is called with the exception message?

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 10]
```diff
@@ -0,0 +1,802 @@
...
+            course_access_configuration_mock.objects
+            .filter.return_value
+            .first.return_value
+        )
+        course_access_conf_queryset_mock.is_course_id_allowed.return_value = False
+
+        with self.assertRaises(LtiToolLaunchException) as ex:
+            self.view_class().check_course_access_permission(COURSE_ID, ISS, AUD)
+        course_access_conf_queryset_mock.is_course_id_allowed.assert_called_once_with(COURSE_ID)
+        self.assertEqual(
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:427]: @alexjmpb Can you validate gettext is called with the exception message?

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 11]
```diff
@@ -0,0 +1,802 @@
...
+        Args:
+            mark_user_change_as_expected_mock: Mocked mark_user_change_as_expected function.
+            login_mock: Mocked login function.
+            authenticate_mock: Mocked authenticate function.
+        """
+        authenticate_mock.return_value = None
+
+        with self.assertRaises(LtiToolLaunchException) as ex:
+            self.view_class().authenticate_and_login(None, **BASE_LAUNCH_DATA)
+        self.assertEqual(
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:520]: @alexjmpb Can you validate gettext is called with the exception message?

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 12]
```diff
@@ -0,0 +1,802 @@
...
+            str(ex.exception),
+            'Profile authentication failed.',
+        )
+        authenticate_mock.assert_called_once_with(None, **BASE_LAUNCH_DATA)
+        login_mock.assert_not_called()
+        mark_user_change_as_expected_mock.assert_not_called()
+
+
+@patch('openedx_lti_tool_plugin.views.course_enrollment')
+class TestLtiToolLaunchViewEnroll(TestLtiToolLaunchViewBase):
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:530]: @alexjmpb Does this cases cover when course_enrollment_exception is raised?

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 13]
```diff
@@ -0,0 +1,802 @@
...
+
+        Args:
+            redirect_mock: Mocked redirect function.
+            allow_complete_course_launch_mock: Mocked 'allow_complete_course_launch' configuration.
+        """
+        allow_complete_course_launch_mock.is_enabled.return_value = False
+
+        with self.assertRaises(LtiToolLaunchException) as ex:
+            self.view_class().get_course_launch_response(COURSE_ID)
+        self.assertEqual(
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:598]: @alexjmpb Can you validate gettext is called with the exception message?

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 14]
```diff
@@ -0,0 +1,802 @@
...
+
+        Args:
+            usage_key_mock: Mocked usage_key object for generating test data.
+            redirect_mock: Mocked redirect function for testing redirection.
+        """
+        usage_key_mock.from_string.return_value = MagicMock(course_key='different-course-id')
+
+        with self.assertRaises(LtiToolLaunchException) as ex:
+            self.view_class().get_unit_component_launch_response(USAGE_KEY, COURSE_ID)
+        self.assertEqual(
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:653]: @alexjmpb Can you validate gettext is called with the exception message?

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 15]
```diff
@@ -0,0 +1,802 @@
...
+            redirect_mock: Mocked redirect function for testing redirection.
+        """
+        usage_key_mock.from_string.return_value = MagicMock(
+            course_key=COURSE_ID,
+            block_type=block_type,
+        )
+
+        with self.assertRaises(LtiToolLaunchException) as ex:
+            self.view_class().get_unit_component_launch_response(USAGE_KEY, COURSE_ID)
+        self.assertEqual(
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:680]: @alexjmpb Can you validate gettext is called with the exception message?

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 16]
```diff
@@ -0,0 +1,802 @@
...
+        }
+
+        with self.assertRaises(LtiToolLaunchException) as ex:
+            self.view_class().handle_ags(
+                launch_message,
+                launch_data,
+                LTI_PROFILE,
+                COURSE_ID,
+            )
+        self.assertEqual(
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:738]: @alexjmpb Can you validate gettext is called with the exception message?

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 17]
```diff
@@ -0,0 +1,802 @@
...
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:765]: @alexjmpb Can you validate gettext is called with the exception message?

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 18]
```diff
@@ -0,0 +1,802 @@
...
+@patch.object(LtiToolLaunchView, 'check_course_access_permission')
+@patch.object(LtiToolLaunchView, 'get_lti_profile')
+@patch.object(LtiToolLaunchView, 'authenticate_and_login')
+@patch.object(LtiToolLaunchView, 'enroll')
+@patch.object(LtiToolLaunchView, 'handle_ags')
+@patch.object(LtiToolLaunchView, 'get_resource_launch', return_value=(MagicMock(), COURSE_KEY))
+@patch('openedx_lti_tool_plugin.views.DjangoDbToolConf')
+@patch('openedx_lti_tool_plugin.views.DjangoCacheDataStorage')
+@patch.object(CourseKey, 'from_string', return_value=COURSE_KEY)
+class TestLtiToolLaunchViewPost(TestLtiToolLaunchViewBase):
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:44]: @alexjmpb There is a test case missing to verify LoggedHttpResponseBadRequest is returned when LtiToolLaunchException is raised

File: openedx_lti_tool_plugin/views.py [hunk 19]
```diff
@@ -252,87 +427,41 @@ def post(
...
 
-            LtiGradedResource.objects.get_or_create(
-                lti_profile=lti_profile,
-                context_key=context_key,
-                lineitem=lineitem,
-            )
+                return resource_response
 
-        return response
+            raise LtiToolLaunchException(_('Only resource link launch request are supported.'))
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/views.py:462]:
```suggestion
            raise LtiToolLaunchException(_('Only resource launch requests are supported.'))
```

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py (hunk 12 above)
Comment (alexjmpb) [reply openedx_lti_tool_plugin/tests/views/test_launch_view.py:600]: No I added one

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 20]
```diff
@@ -0,0 +1,888 @@
...
+        self.assertEqual(
+            self.view_class().get_identity_claims(launch_data_mock),
+            (
+                launch_data_mock.get('iss'),
+                get_client_id_mock.return_value,
+                launch_data_mock.get('sub'),
+                get_pii_from_claims_mock.return_value,
+            )
+        )
+
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:338]: Unnecessary blank line. @alexjmpb

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 21]
```diff
@@ -0,0 +1,888 @@
...
+        save_pii_data_mock.is_enabled.return_value = True
+
+        self.assertEqual(
+            self.view_class().get_identity_claims(launch_data_mock),
+            (
+                launch_data_mock.get('iss'),
+                get_client_id_mock.return_value,
+                launch_data_mock.get('sub'),
+                get_pii_from_claims_mock.return_value,
+            )
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:336]: Trailing comma. @alexjmpb

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 22]
```diff
@@ -0,0 +1,888 @@
...
+        save_pii_data_mock.is_enabled.return_value = False
+
+        self.assertEqual(
+            self.view_class().get_identity_claims(launch_data_mock),
+            (
+                launch_data_mock.get('iss'),
+                get_client_id_mock.return_value,
+                launch_data_mock.get('sub'),
+                {},
+            )
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:370]: Trailing comma. @alexjmpb

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 23]
```diff
@@ -0,0 +1,888 @@
...
+            redirect_mock: Mocked redirect function for testing redirection.
+        """
+        usage_key_mock.from_string.return_value = MagicMock(course_key='different-course-id')
+
+        with self.assertRaises(LtiToolLaunchException):
+            self.view_class().get_unit_component_launch_response(USAGE_KEY, COURSE_ID)
+        gettext_mock.assert_called_once_with('Unit/component does not belong to course.')
+        redirect_mock.assert_not_called()
+
+    @data('sequential', 'chapter')
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:747]: Please, add 'course' case. @alexjmpb

File: openedx_lti_tool_plugin/views.py [hunk 24]
```diff
@@ -225,20 +288,132 @@ def enroll(self, request: HttpRequest, edx_user: UserT, course_key: str):
...
+            lti_profile: LTI profile associated to the launch claims.
+            context_key: Course ID or Unit/component key string.
+
+        Returns:
+            Instance of LtiGradedResource.
+
+        Raises:
+            LtiToolLaunchException when lineitem or AGS scope are missing.
+        """
+        if not launch_message.has_ags():
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/views.py:365]: Please, add a test case for this scenario. @alexjmpb

File: openedx_lti_tool_plugin/views.py [hunk 25]
```diff
@@ -225,20 +288,132 @@ def enroll(self, request: HttpRequest, edx_user: UserT, course_key: str):
...
+        ags_endpoint = launch_data.get(AGS_CLAIM_ENDPOINT, {})
+        lineitem = ags_endpoint.get('lineitem')
+
+        if not lineitem:
+            raise LtiToolLaunchException(_('Missing AGS lineitem.'))
+
+        if AGS_SCORE_SCOPE not in ags_endpoint.get('scope', []):
+            raise LtiToolLaunchException(_(f'Missing required AGS scope: {AGS_SCORE_SCOPE}'))
+
+        return LtiGradedResource.objects.get_or_create(
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/views.py:377]: Missing test case for this. @alexjmpb

**Review by Squirrel18 — CHANGES_REQUESTED** [review]

**Review by Squirrel18 — APPROVED** [review]

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 26]
```diff
@@ -0,0 +1,919 @@
...
+                call('Only resource launch requests are supported.'),
+                call(f'LTI 1.3 Launch failed: {gettext_mock.return_value}'),
+            ],
+            any_order=True,
+        )
+        self.assertEqual(response.status_code, 400)
+
+    @patch('openedx_lti_tool_plugin.views.LoggedHttpResponseBadRequest')
+    @patch('openedx_lti_tool_plugin.views._')
+    def test_post_returns_404_when_launch_exception_raised(
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:234]:
```suggestion
    def test_post_returns_400_when_launch_exception_raised(
```

File: openedx_lti_tool_plugin/tests/views/test_launch_view.py [hunk 27]
```diff
@@ -0,0 +1,919 @@
...
+        handle_ags_mock: MagicMock,  # pylint: disable=unused-argument
+        enroll_mock: MagicMock,  # pylint: disable=unused-argument
+        authenticate_and_login_mock: MagicMock,  # pylint: disable=unused-argument
+        get_lti_profile_mock: MagicMock,  # pylint: disable=unused-argument
+        check_course_access_permission_mock: MagicMock,  # pylint: disable=unused-argument
+        get_identity_claims_mock: MagicMock,  # pylint: disable=unused-argument
+        get_launch_data_mock: MagicMock,  # pylint: disable=unused-argument
+        message_launch_mock: MagicMock,
+    ):
+        """Test POST request without a resource launch.
```
Comment (kuipumu) [inline openedx_lti_tool_plugin/tests/views/test_launch_view.py:251]: Please update docstring to reflect test case

**Pull request merged**